### PR TITLE
Adjust Webpack config to improve rebuild time on dev environments

### PR DESCRIPTION
**Pull request by jmkulwik**
In some webpack config testing, we happened upon a discovery improves rebuild time by as much as 90%. This improvement does come at a cost - the better rebuild increases the page load time. However, when page load is taken into account, the rebuild is still as much as 75% better on many pages and 40% better on high load pages such as Java Lab levels.

The discovery: Removing both the splitChunks optimization and the StatsWriterPlugin significantly decreases the build time. 

### Stats Writer Plugin

The [StatsWriterPlugin](https://github.com/FormidableLabs/webpack-stats-plugin?tab=readme-ov-file) I believe is being used to track out bundle sizes in our WebPack build. Evidence for this is from the [PR that introduced it](https://github.com/code-dot-org/code-dot-org/pull/24440). This will be a useful tool as we work on breaking up our WebPack build, however, it adds between 9 and 13 seconds to every build and rebuild (see table below).

Removing the StatsWriterPlugin from builds on the dev environment should have no negative effect. It is useful when specifically analyzing bundle sizes, but is not necessary for work unrelated to build optimization. Individuals working on build optimization can choose to re-enable this for their local builds without other developers incurring the increased build time cost.

### Split Chunks Optimization

The [splitChunks optimization](https://webpack.js.org/configuration/optimization/#optimizationsplitchunks) allows us to customize the WebPack chunking. This improves our page load time for our end users and is important to keep in our production and CI environments. However, the optimized chunking increases the rebuild time by as much as 5x (see table below).

Removing the splitChunks optimization from builds on dev environments will also improve the development experience in the vast majority of cases. If, however, an individual is rebuilding rarely and is working on page navigation on high load pages, such as Java Lab levels, this may increase development time (see table below). NOTE: THIS SHOULD NOT BE REMOVED FROM OUR PRODUCTION BUILDS

_Implementation Note: disabling StatsWriterPlugin always improves build time. However, disabling splitChunks only improves build time if StatsWriterPlugin is also disabled - otherwise the removal of splitChunks causes the StatsWriterPlugin to dramatically increase its runtime, leading to far longer build times than had the splitChunks step run as well._

### Performance Comparisons

The below table shows the improvements in rebuild-and-load time on various pages on a benchmark environment with consistent background processes running. Specifically, a developer has already run `yarn start` to automatically rebuild changes. Then, a developer changes a file and saves it and rebuild is automatically kicked off (`rebuild` step below). After the rebuild, the developer reloads the page (`reload` step below). The sum of these two steps is the `total` time below.

Edit: after an ask for more detail, I've added another item to the tables below, `refresh`. This is the time it takes to refresh the page with no code change (i.e. you've already done the `reload` step and hit the refresh button again).

For the sake of readability, "StatsWriterPlugin" is shortened to "Stats" and "splitChunks" is shortened to "Chunks." There is a few second variability in identical executions of the same step.

On a Java Lab level page:
| Code Change | Rebuild | Reload | Total | _Refresh_ |
| --------------- | -------- | -------- | ----- | -------- |
| None | 30s | 7s | 37s | _4s_ |
| no Stats | 21s | 6s | 27s | |
| no Stats or Chunks | 4s | 16s | 20s | _4s_ |

On a Project Beats (Lab2) project page:
| Code Change | Rebuild | Reload | Total | _Refresh_ |
| --------------- | -------- | -------- | ----- | -------- |
| None | 33s | 5s | 38s | _3s_ |
| no Stats | 20s | 5s | 25s | |
| no Stats or Chunks | 4s | 8s | 12s | _3s_ |

On a logged out studio.code.org page:
| Code Change | Rebuild | Reload | Total | _Refresh_ |
| --------------- | -------- | -------- | ----- | -------- |
| None | 33s | 4s | 36s | _3s_ |
| no Stats | 21s | 3s | 24s | |
| no Stats or Chunks | 4s | 5s | 9s | _3s_ |

Finally, the initial build time also improves. One benchmark clocked the initial `yarn start` build at 86s. It decreased to 72s without Stats and decreased further to 48s without Stats or Chunks.

### Affect to other environments

There should be no affect to non-dev environments. The build.rake file shows that the build step will conditionally `build:dist` or `build`, depending on the `optimize_webpack_assets` CDO flag. This is set to true in the default config.yml.erb file and is only overridden in the development.yml.erb file. The `build` step as specified in the package.json has the flag `DEV=1`, which is what we depend on with this code change. The `build:dist` step does not have this flag. Testing locally with this change, the `build:dist` command does run the Stats and Chunks steps and the `build` step does not.

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/webpack.config.js [hunk 1]
```diff
@@ -418,108 +418,108 @@ function createWebpackConfig({
...
+      //     p5lab: {
+      //       name: 'p5-dependencies',
+      //       priority: 10,
+      //       minChunks: 2,
+      //       chunks: chunk =>
+      //         ['spritelab', 'gamelab', 'dance'].includes(chunk.name),
+      //       test: module => /p5/.test(module.resource),
+      //     },
+      //   },
+      // },
```
Comment (davidsbailey) [inline apps/webpack.config.js]: 👋 Vijaya mentioned you were soliciting early feedback on this PR. does commenting out this block help local development times? my impression was that splitChunks only affected the optimized/production build.

Comment (jmkulwik) [reply apps/webpack.config.js]: Thanks Dave. I updated the description with timing benchmarks - not sure whether something changed since we first added splitChunks, but at present, it does appear to affect the dev build.

Comment (davidsbailey) [reply apps/webpack.config.js]: I see, nice find! enabling this only outside of DEV mode makes sense to me

**Review by bencodeorg — APPROVED** [review]
This is awesome and makes sense to me! Two things I was thinking about:
1. is there any risk of by not chunking in development, we'll have some diff in behavior between development and other environments? I think the answer is no.
2. The calculations you list assume one page reload per rebuild, which I think is a reasonable expectation. But worth noting the tradeoff to exclude splitChunks is less appealing if you're doing multiple page reloads per build (ie, navigating around the site, etc)

**Comment by davidsbailey** [comment]
> This is awesome and makes sense to me! Two things I was thinking about:
> 
> 1. is there any risk of by not chunking in development, we'll have some diff in behavior between development and other environments? I think the answer is no.

correct -- there is already a big difference between js in production vs development, and this shouldn't make the gap any wider. if you get a problem that repros in prod only, there are ways to get your local environment to run like prod. https://github.com/code-dot-org/code-dot-org/blob/staging/apps/docs/build.md

**Comment by bencodeorg** [comment]
> > 1. is there any risk of by not chunking in development, we'll have some diff in behavior between development and other environments? I think the answer is no.
> 
> correct -- there is already a big difference between js in production vs development, and this shouldn't make the gap any wider. if you get a problem that repros in prod only, there are ways to get your local environment to run like prod. https://github.com/code-dot-org/code-dot-org/blob/staging/apps/docs/build.md

Thanks, yeah I meant like a new problematic difference. Seems like not!

**Review by sanchitmalhotra126 — APPROVED** [review]
Pulled this locally and worked with this for some time today, and it works great! Thanks for doing this!

File: apps/webpack.config.js [hunk 2]
```diff
@@ -615,6 +510,117 @@ function createWebpackConfig({
     ],
   };
 
+  // These two steps increase the rebuild time by 10x and are not required on dev builds. Exclude
+  // these in dev environments, where rebuild is a very common action.
```
Comment (snickell) [inline apps/webpack.config.js]: Appreciate this comment, betting posterity will too

**Review by snickell — COMMENTED** [review]
I'm potentially having some problems with this branch, hoping/guessing its something locally wrong. 

I merged this branch with `unfirebase`, and restarted `yarn start`. When I tried to visit an applab project (say: http://localhost-studio.code.org:3000/projects/applab/Yp05MnSdVubn04tBoEZn_g/edit ), I got no obvious errors in the JS console, the page didn't load beyond a spinning wheel of wait.
image

When switching back the _unmerged_ working branch (`unfirebase`), and restarting `yarn start` again, the page loads into applab as expected. I've switched back and forth a few times, with the pattern consistent.

Could this be something caching related since I'd already loaded that exact URL with the old webpack config? @sanchitmalhotra126 / @jmkulwik does applab work for you with this webpack?

That said, build speed was ⚡ ⚡ ⚡ ! I'll try to dig deeper tomorrow, but I also think if somebody else has no problem switching back and forth on an applab project, I might not read too much into this?

**Comment by bencodeorg** [comment]
@snickell I can repro locally in applab. I have the same issue with a new applab project, so I don't think it's a caching thing?

**Comment by sanchitmalhotra126** [comment]
@snickell I can also repro locally in applab. I noticed that the `levels-applab-main.js` entrypoint is not present when looking at sources in dev tools for some reason. I'm not sure why; I tried a few other level types (Dance, Sprite Lab, Game Lab, Artist) and I don't see the same issue.

**Comment by snickell** [comment]
@sanchitmalhotra126 curiously, @jmkulwik was reporting loading problems with spritelab

Ironically, I think a good tool to debug this is to temporarily re-enable `StatsWriterPlugin` on dev 🤭

File: apps/webpack.config.js [hunk 3]
```diff
@@ -615,6 +510,117 @@ function createWebpackConfig({
...
+  if (!process.env.DEV) {
+    WEBPACK_CONFIG.optimization.splitChunks = {
```
Comment (snickell) [inline apps/webpack.config.js]: I waffled a bit on this when I first structured `webpack.config.js`, but I'd prefer to continue the style of the file not imperatively manipulating WEBPACK_CONFIG after its declaration.

This is part of what led its predecessor to being very confusing. Its a tradeoff because imperative syntax is simpler, but I think long-term it results in config that you have to "read as code" in a way that's hard to reason about.

Rather than:
```
if (!process.env.DEV) {
 WEBPACK_CONFIG.optimization.splitChunks = {
    // SPLIT CHUNKS CONFIG GOES HERE
  }
}
```

Do the declaration inline to WEBPACK_CONFIG using a ternary operator to add the conditionality:
```
const WEBPACK_CONFIG = {
    // ...
    optimization: {
      splitChunks: process.env.DEV ? {
        // SPLIT CHUNKS CONFIG GOES HERE
      } : undefined,
    }
}
```

Comment (snickell) [reply apps/webpack.config.js]: I've made the modifications I would suggest on branch `jmkulwik/speed-up-builds-reformatted`, PR: https://github.com/code-dot-org/code-dot-org/pull/55836 . If @jmkulwik agrees, can just click to merge merge https://github.com/code-dot-org/code-dot-org/pull/55836 into this one (its setup to merge to this PR not staging).

I know this style looks more complex, and the ternary here isn't pretty, but it results in complex config that can be read linearly (albeit with lots of line noise). The model is "templated config file".

**Review by snickell — CHANGES_REQUESTED** [review]
I'd prefer if `webpack.config.js` continued to use a templated-config type approach, rather than imperatively modifying the config after its declared. The imperative approach is simpler for adding one or two conditionals, but we've already got a dozen of them, and unwinding the previously added imperative conditions was a big part of my webpack cleanup a several months ago.

I made a PR that merges into this one that does the fix, if @jmkulwik agrees: https://github.com/code-dot-org/code-dot-org/pull/55836

**Comment by snickell** [comment]
I just diffed the JS files loaded from network for before/after this PR on applab, and its loading the same files (modulo a couple calls to Google JS apis, which I assume are from analytics and unrelated).

TL;DR: working and non-working cases load the same files, so this is not a case of needing to re-configure our rails asset manager to deal with different generated JS entrypoints, etc.

network-load-diff

@sanchitmalhotra126 I'm guessing `levels-applab-main.js` in the devtools sources is a sourcemap from a larger file like `code-studio.js`?

This is all rather odd, because I would find the reverse situation more likely (chunks aren't loading).... it should be really simple when there are no chunks and everything is bundled together 🤔

**Comment by snickell** [comment]
I used the Chrome devtools JS profiler to look for where execution differs between working and not-working applab loading cases.

TL;DR: execution of `applab.js:1` fails early in the script evaluation phase in the not-working case, what follows will be highlighting differences between the working and not-working profiles as we zoom in closer and closer on what I was calling `Callstack #1` (its applab.js:1)

profile-missing-calls

Top profile is a working applab, bottom is the non-working applab.

I circled and linked 4 key differences between them, which visually appear to be a direct chain of cause and effect:
1. `Callstack #1` never executes
2. Therefore `main.json is never loaded`
3. Therefore `Callstack #2` never executes after it loads
4. Therefore page never switches to rendered condition and stays at the spinner.

I zoomed in on `Callstack #1` so you can see more precisely which block is not running:
zoomed-on-first-callstack

Finally, zooming farther on the working `Callstack #1` to see what it is, we find that its beginning evaluation, loading and execution of `applab.js:1`:
callstack-1-is-loading-applab js

Note that `applab.js` is a "real file" (i.e. its not a virtual file inferred by a source map, its actually transferred across the network). `applab.js` is present in the not-working condition too, so its not just a missing file. I set a breakpoint in the not-working condition on `applab.js:1`, and it does trigger.

....but that's where I got stuck using a profiling based approach :-P

**Comment by snickell** [comment]
One final thing I noticed while reading through `webpack.config.js` for clues (see: https://github.com/code-dot-org/code-dot-org/blob/ffcfd387da24d5872b9ec69a2edfcedfaf5d358f/apps/webpack.config.js/#L451-L471 for the full comment)
```
       // SNIP, there's more comment above here, but I removed it for emphasis.
 
      // Without a single, named runtimeChunk there would be no runtimeChunk
      // to include, and entry points would load and run separately.
      // However, those entry points would create separate instances of any
      // shared modules. This would mean that state within webpack modules
      // cannot be shared between entry points, breaking many assumptions made
      // by our application. SNIP
      //
      // In the future, if we can limit ourselves to one webpack entry point
      // per page, we could consider removing the runtimeChunk config.
      runtimeChunk: {
        name: 'webpack-runtime',
      },
```

Is it possible that `applab` has multiple entry points, and we interfered with named runtime chunks by virtue of disabling splitChunks? The comment suggests that some sort of bad not-state-sharing might result.... maybe this is a result of that?

**Comment by davidsbailey** [comment]
I've been loosely following along here, and one thought is that it might be possible to recoup most of the performance gains by providing a much simpler splitChunks config, rather than trying to remove it entirely.

another idea here is to try temporarily setting `optimize_webpack_assets: false` in ui_tests.sh, like we do here: https://github.com/code-dot-org/code-dot-org/blob/e802088401b4913415bd6db951e244acc73a205d/docker/unit_tests.sh#L32 this might take some experimentation, but the idea would be to leverage our existing UI test coverage to convince yourself that you haven't broken any part of our web site when running in local development. for example, I would expect any test using applab to fail if you tried this on the current PR.

**Comment by snickell** [comment]
Tying back to @sanchitmalhotra126 noticing that `levels-applab-main.js` was missing (I believe from the sourcemaps in the devtool sources tab)... you can see that the not-working applab.js does clearly include `levels-applab-main.js` as its primary entry point code. The problem is that even though the file itself loaded, its default webpack entrypoint is never triggered.

I inspected the webpack generated applab.js for both working and not working cases. This is the end of both files, where webpack puts the code it loads by default:
applab js

Curiously while both working and not-working load `applab.js:1` (breakpoint hits on line 1), the "main" `__webpack_require__` entrypoint (bottom of both files) only hits the breakpoint in the working case.

File: apps/webpack.config.js (hunk 3 above)
Comment (jmkulwik) [reply apps/webpack.config.js]: Thanks Seth! I definitely like the ternary better than the conditionals. I've merged it in.

**Review by snickell — CHANGES_REQUESTED** [review]
@bencodeorg figured out that we had stale files like `code-studio-common.js` in our `apps/build` dirs generated by splitChunks and still referenced by, e.g. `application.html.haml`. So we were still loading old chunks, and they were obviously conflicting with the new chunks. Cool!

So to get this merged the issues we know of are:
1. Warn that you need to do a `yarn clean` if building with DEV=1 and we see that there's still a `code-studio-common.js` in `apps/build`
2. Remove links to `code-studio-common.js` and kin from `application.html.haml` and kin, when running Dashboard in RAILS_ENV=dev so we don't get 404s. There might be some detail to think through here because I believe its possible and potentially (???) sane to run `yarn start` in DEV mode, but have Dashboard running in testing.

**Comment by bencodeorg** [comment]
@snickell @jmkulwik here's a PR to potentially merge into this branch to mitigate the first issue Seth mentioned above:

https://github.com/code-dot-org/code-dot-org/pull/55891

**Comment by sanchitmalhotra126** [comment]
@snickell @bencodeorg nice find! I was also able to run App Lab successfully after running `yarn clean` 🎉 any ideas why the conflicting chunks were affecting only App Lab in particular?

**Comment by bencodeorg** [comment]
@sanchitmalhotra126 I have no idea :)

**Comment by davidsbailey** [comment]
you could try `yarn build:analyze` to find out!

File: apps/webpack.config.js [hunk 4]
```diff
@@ -601,6 +603,15 @@ function createWebpackConfig({
           }
           return file;
         },
+        ...(process.env.DEV
```
Comment (jmkulwik) [inline apps/webpack.config.js:606]: Thinking about how we're using `CDO.optimize_webpack_assets` on the dashboard side of things, would it make sense to use the same CDO flag here? (Trying to remember what the webpack.config has access to.)

Comment (jmkulwik) [reply apps/webpack.config.js:606]: @bencodeorg ^ Referring to your new PR https://github.com/code-dot-org/code-dot-org/pull/55925/files

Comment (bencodeorg) [reply apps/webpack.config.js:606]: Yeah I was wondering about that...I was thinking that wouldn't work in a non-Rails context (and it doesn't appear we currently can access that sort of config easily in `apps` code)? I could be totally wrong though. We could add another flag to `start`, `build`, etc like `OPTIMIZE_WEBPACK_ASSETS=1` if someone is doing a string search at least? Or that might be more confusing 😬

Comment (snickell) [reply apps/webpack.config.js:606]: I personally wouldn't want to use/reference CDO in JS build bits. As @bencodeorg mentions, we'd have to build the bridge, but moreover I don't think its a bridge that should be built.

Comment (snickell) [reply apps/webpack.config.js:606]: Crazy idea, possibly bad.... what if we sniff if the file exists from the rails side too, instead of indexing off a flag (`CDO.optimize_webpack_assets`)?

```
- if file_exists('js/code-studio-common.js') || CDO.optimize_webpack_assets
      %script{src: webpack_asset_path('js/vendors.js')}
      %script{src: webpack_asset_path('js/code-studio-common.js')}
```

It /probably/ doesn't matter if we add an extra fs syscall or two to the default page template, but it'd also be really easy to memoize file_exists in production mode (in dev mode you want it to always check so you don't have to restart `dashboard-server` if you change JS build type).

Comment (snickell) [reply apps/webpack.config.js:606]: Possibly the right logic would be `if file_exists('js/code-studio-common.js') || CDO.optimize_webpack_assets`, which would give you 404s if you had done the wrong JS build type for current prod-like environment.

Comment (bencodeorg) [reply apps/webpack.config.js:606]: Are you thinking it's a performance or just general code improvement to do this by checking for existence of the custom webpack chunk files? I was wondering if we have any existing usages of `CDO.optimize_webpack_assets` that gets hit as part of our production code, looks like any time `webpack_asset_path` is called that config value is checked:

https://github.com/code-dot-org/code-dot-org/blob/59fc56e18376beeda7df10952b7b456604544355/lib/cdo/asset_helper.rb#L35-L43

I think I find what I have here a bit clearer than checking whether one of those webpack files exists, but open to changing if if you're feeling like it's a better direction.

Comment (snickell) [reply apps/webpack.config.js:606]: My thinking is, that I'm not personally confident that DEV(js)==CDO.optimize_webpack_assets(rails) in all conditions. It could well be true, but it requires a lot of knowledge to reason about that I don't have. On the other hand, by looking at the filesystem (shared between webpack node and rails), we know that both systems are indexing off the same state, so you can't have a state tear.

I'm not worried about the performance of `CDO.optimize_webpack_assets`, I'm worried that it might not have a one-to-one correspondence with DEV=1 being set during the yarn build.

I /was/ slightly worried that file_exists on every template render could conceivably slow things down, but rails is probably doing so many filesystem calls that its lost in the noise (and I suggested an optimization for that). But the optimization isn't relevant unless we thought file_exists was a more consistent/better check to match DEV than the CDO flag.

Comment (davidsbailey) [reply apps/webpack.config.js:606]: It seems totally reasonable to me to set DEV=1 if (and only if) `!CDO.optimize_webpack_assets`, and we could do it via similar mechanism to `generateSharedConstants.rb`. However I am also not seeing a need for this -- 99% of the time the developer wants to run with DEV=1, and if they want to run with optimized JS they will be consulting special instructions and it will be easy enough to set optimized js there "manually" in that case.

or am i missing a new, increased need for running optimized js locally?

Comment (bencodeorg) [reply apps/webpack.config.js:606]: I'm unaware of a new need to run optimized JS locally more regularly :). I think mostly it's that we (especially me) are mucking around in code we don't fully understand, so wanted to be cognizant the implications of any changes we're making.

I feel more confident having your input that this seems reasonable though Dave, as someone who is very familiar with our JS build process.

I'm kind of inclined to get this merged (including the two branch off branch PRs) and out to folks and see if we run into any issues -- thoughts?

Comment (snickell) [reply apps/webpack.config.js:606]: Yeah, I felt the same as Ben, we know how much we don't know! I think Dave weighing in here is the kicker, and it makes sense to merge, get it into folk's hands, and handle any future changes as a followup (if anything does crop up).

**Review by snickell — APPROVED** [review]
Ready to merge™️
